File: 1130-last-stone-weight-ii/last-stone-weight-ii.cpp
class Solution {
    int OFFSET = 3000;
    int dp[32][6004];
public:
    Solution(){
        memset(dp, -1, sizeof(dp));
    }
    int solve(vector<int> &nums, int n, int sum){
        if (n == nums.size()){
            if (sum < 0) return 5000;
            return sum;
        }
        
        if (dp[n][OFFSET+sum] != -1) return dp[n][OFFSET+sum];

        int add = solve(nums, n+1, sum + nums[n]);
        int sub = solve(nums, n+1, sum - nums[n]);

        dp[n][OFFSET+sum] = min(add, sub);
        
        return dp[n][OFFSET+sum];
    }
    int lastStoneWeightII(vector<int>& stones) {
        return solve(stones, 0, 0);
    }
};

File: 1130-last-stone-weight-ii/last-stone-weight-ii.go
const OFFSET = 3001

func solve(stones []int, idx int, sum int, total int, dp [][]int) int {
    if idx >= len(stones) {
        return abs(total - 2 * sum)
    }

    if dp[idx][sum + OFFSET] != -1 {
        return dp[idx][sum + OFFSET]
    }

    take := solve(stones, idx+1, sum + stones[idx], total, dp)
    not_take := solve(stones, idx+1, sum, total, dp)

    dp[idx][sum + OFFSET] = min(take, not_take)

    return dp[idx][sum + OFFSET]
}
func lastStoneWeightII(stones []int) int {
    n := len(stones)
    dp := make([][]int, n)
    for i := range dp {
        dp[i] = make([]int, 2 * OFFSET)
        for j := range dp[i] {
            dp[i][j] = -1
        }
    }

    total := 0
    for _, num := range stones {
        total += num
    }
    return solve(stones, 0, 0, total, dp)
}
func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}